File: leetcode/dia 042/numIslands.py
class Solution:
    def numIslands(self, grid: List[List[str]]) -> int:
        def search(s, i, j):
                if (i, j) in s or grid[i][j] == "0":
                    return False
                s.add((i, j))
                for m in range(i - 1, i + 2):
                    for n in range(j - 1, j + 2):
                        if m - i != n - j and m in range(0, len(grid)) and n in range(0, len(grid[0])) and (m == i or n == j):
                            search(s, m, n)
                return True
        s = set()
        total = 0
        for i in range(len(grid)):
            for j in range(len(grid[i])):
                if grid[i][j] != "0" and search(s, i, j):
                    total += 1
        return total
        

File: leetcode/dia 042/numIslands.js
var numIslands = function(grid) {
    function explore(set, i, j){
        if (set.has(i+','+j) || grid[i][j] === "0") return false;
        set.add(i+','+j);
        if (i - 1 >= 0) explore(set, i - 1, j);             
        if (i + 1 < grid.length) explore(set, i + 1, j);        
        if (j - 1 >= 0) explore(set, i, j - 1);             
        if (j + 1 < grid[0].length) explore(set, i, j + 1);        
        return true;
    }
    let total = 0, set = new Set();
    for (let i = 0; i < grid.length; i++){
        for (let j = 0; j < grid[0].length; j++){
            if (explore(set, i, j)) total += 1;   
        }
    }
    return total;
};